timemachine/fe/free_energy.py

```diff
@@ -1259,6 +1243,67 @@ def compute_u_kn(trajs, initial_states) -> Tuple[NDArray, NDArray]:
     return u_kn, np.array(N_k)
 
 
+def generate_pair_bar_ulkns(
+    initial_states: Sequence[InitialState],
+    samples_by_state: Sequence[Trajectory],
+    temperature: float,
+    unbound_impls: Sequence[custom_ops.Potential] | None,
+) -> NDArray:
+    """Generate pair bair u_klns.
+    This is a specialized variant of generating u_klns, only loading each set of frames into memory once.
+    Each set of frames is loaded once then all of the parameters of interest are run in a batch.
+    This improves throughput for potentials that use Neighborlists, as there are at most len(frames) neighborlist
+    rebuilds, rather than 3 * len(frames).
+
+    Returns
+    -------
+        u_klns: np.array[len(initial_states) - 1, len(unbound_impls), 2, 2, n_frames]
+    """
+
+    assert len(initial_states) > 0
+    assert len(initial_states) == len(samples_by_state)
+    if unbound_impls is None:
+        unbound_impls = [pot.potential.to_gpu(np.float32).unbound_impl for pot in initial_states[0].potentials]
+    assert len(unbound_impls) == len(initial_states[0].potentials)
+    kBT = temperature * BOLTZ
+    # Construct an empty array
+    state_to_params = np.zeros((len(initial_states), len(initial_states), len(unbound_impls)), dtype=object)
+    for i, state in enumerate(initial_states):
+        frames = np.array(samples_by_state[i].frames)
+        boxes = np.asarray(samples_by_state[i].boxes)
+
+        state_idxs = []
+        if i > 0:
+            state_idxs.append(i - 1)
+        state_idxs.append(i)
+        if i < len(initial_states) - 1:
+            state_idxs.append(i + 1)
+        for j, pot in enumerate(state.potentials):
+            params = np.array([initial_states[idx].potentials[j].params for idx in state_idxs])
+            _, _, Us = unbound_impls[j].execute_batch(
```

Comment:
Would it be possible to use `execute_batch_sparse` here instead (similar to the usage in [`compute_potential_matrix`](https://github.com/proteneer/timemachine/blob/caaaad2a43756435b18e7303f2621ccb7220d4db/timemachine/fe/free_energy.py#L1116)) to compute the full `U_kln` matrix with a single custom_ops call per potential?

Reply:
Never mind - I wasn't quite understanding what this was doing. I don't see an obvious way of doing this.